src/server/debug.rs

```diff
@@ -444,6 +482,223 @@ impl Debugger {
+            write_fix_count: 0,
+        })
+    }
+
+    fn min_key(key: Option<Vec<u8>>, iter: &DBIterator, f: fn(&[u8]) -> &[u8]) -> Option<Vec<u8>> {
+        let iter_key = if iter.valid() {
+            Some(f(keys::origin_key(iter.key())).to_vec())
+        } else {
+            None
+        };
+        match (key, iter_key) {
+            (Some(a), Some(b)) => if a < b {
+                Some(a)
+            } else {
+                Some(b)
+            },
+            (Some(a), None) => Some(a),
+            (None, Some(b)) => Some(b),
+            (None, None) => None,
+        }
+    }
+
+    pub fn check_mvcc(&mut self, wb: &WriteBatch) -> Result<()> {
+        loop {
+            // Find min key in the 3 CFs.
+            let mut key = MvccChecker::min_key(None, &self.default_iter, truncate_ts);
+            key = MvccChecker::min_key(key, &self.lock_iter, |k| k);
+            key = MvccChecker::min_key(key, &self.write_iter, truncate_ts);
+
+            match key {
+                Some(key) => self.check_mvcc_key(wb, key.as_ref())?,
+                None => return Ok(()),
+            }
+        }
+    }
+
+    fn check_mvcc_key(&mut self, wb: &WriteBatch, key: &[u8]) -> Result<()> {
+        let (mut default, mut write, mut lock) = (None, None, None);
+        let (mut next_default, mut next_write, mut next_lock) = (true, true, true);
+        loop {
+            if next_default {
+                default = self.next_default(key)?;
+                next_default = false;
+            }
+            if next_write {
+                write = self.next_write(key)?;
+                next_write = false;
+            }
+            if next_lock {
+                lock = self.next_lock(key)?;
+                next_lock = false;
+            }
+
+            // If lock exists, check whether the records in DEFAULT and WRITE
+            // match it.
+            if let Some(ref l) = lock {
+                // All write records' ts should be less than lock's ts.
+                if let Some((commit_ts, _)) = write {
+                    if l.ts <= commit_ts {
+                        warn!(
+                            "LOCK ts is less than WRITE ts, key: {}, lock_ts: {}, commit_ts: {}",
+                            escape(key),
+                            l.ts,
+                            commit_ts
+                        );
+                        self.delete(wb, CF_LOCK, key, None)?;
+                        self.lock_fix_count += 1;
+                        next_lock = true;
+                        continue;
+                    }
+                }
+
+                // If the lock's type is PUT and contains no short value, there
+                // should be a corresponding default record.
+                if l.lock_type == LockType::Put && l.short_value.is_none() {
+                    match default {
+                        Some(start_ts) if start_ts == l.ts => {
+                            default = self.next_default(key)?;
```

Comment:
Set `next_default = true` here to keep it consistent?

Reply:
Good idea.